### PR TITLE
[megatron] fix: preserve router replay fallback config flag

### PR DESCRIPTION
**Pull request by tntnnlrw**
### What does this PR do?

Preserve the router replay fallback feature flag when Megatron config construction filters override kwargs with `hasattr(cls, key)`.

`apply_router_replay_patch()` now patches both `TransformerConfig` and `MLATransformerConfig`, exposes a class-level `enable_routing_replay = False` default for the converter filter, and still wraps config constructors so the flag can be accepted even when native Megatron versions do not define it.

### Root cause

`check_and_construct_configs()` removes override keys that are not visible as class attributes on the target Megatron config class. The router replay patch previously adjusted the constructor signature, but it did not expose `enable_routing_replay` at the class level and did not cover `MLATransformerConfig`, so the fallback flag could be silently dropped before construction.

### Test

- `PYTHONPATH=. pytest tests/utils/megatron/test_router_replay_patch_config_on_cpu.py -q`
- `python -m ruff check verl/utils/megatron/router_replay_patch.py tests/utils/megatron/test_router_replay_patch_config_on_cpu.py`
- `python -m py_compile verl/utils/megatron/router_replay_patch.py tests/utils/megatron/test_router_replay_patch_config_on_cpu.py`
- `git diff --check HEAD~1..HEAD`
- GitHub CI: `cpu_unit_tests`, `pre-commit`, `sanity`, `type-coverage-check`, `doc_test`, title/secrets checks pass.

### API and Usage Example

No user-facing API or usage changes. This only preserves the existing router replay fallback config path for Megatron versions that do not expose a native `moe_enable_routing_replay` field.

### Design & Code Changes

- Preserve `enable_routing_replay` through verl Megatron config filtering by exposing it as a class-level fallback attribute.
- Apply the fallback patch to both `TransformerConfig` and `MLATransformerConfig`.
- Add CPU-only regression coverage for the fallback path, without adding internal end-to-end scripts.

### Checklist Before Starting

- [x] Searched similar PRs/issues: https://github.com/verl-project/verl/pulls?q=router+replay and related #6647 / #5884.
- [x] PR title follows the verl format: `[megatron] fix: ...`.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Applied pre-commit/lint checks: GitHub `pre-commit` passes; focused local `ruff`, `py_compile`, and whitespace checks also pass.
- [x] Documentation is not updated because this fixes internal Megatron config plumbing and does not change user-facing APIs.
- [x] Added CPU-only unit regression coverage under `tests/utils/megatron/`; it is covered by the existing `cpu_unit_tests` workflow.
- [x] CI has already run on GitHub. Core relevant checks pass; remaining red jobs appear unrelated to this patch path and are summarized in the PR discussion.
- [x] Recipe submodule is not affected.

### Notes

This keeps the PR scoped to the core fix and a CPU-only focused regression test. Internal end-to-end R3 smoke scripts are intentionally not included.

### CODE REVIEW

File: verl/utils/megatron/router_replay_patch.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
     warnings.warn("NPU not support router replay for now.", stacklevel=2)
     MoEAlltoAllTokenDispatcher = None
 from megatron.core.transformer.moe.router import TopKRouter
-from megatron.core.transformer.transformer_config import TransformerConfig
+from megatron.core.transformer import MLATransformerConfig, TransformerConfig
```
[the remark this answers is not in the document]
Comment (tntnnlrw) [reply verl/utils/megatron/router_replay_patch.py]: Fixed in 0a7d3827. `MLATransformerConfig` is now imported conditionally so older Megatron-LM versions without MLA config can still import this module.

File: verl/utils/megatron/router_replay_patch.py [hunk 2]
```diff
@@ -337,62 +337,58 @@ def patched_routing(self, logits: torch.Tensor, *args, **kwargs):
...
+
+            config_cls.__init__ = patched_config_init
+            config_cls._verl_router_patched = True
+
+        # config_converter filters kwargs with hasattr(cls, key), so expose a class-level default.
+        if "enable_routing_replay" not in config_cls.__dict__:
+            config_cls.enable_routing_replay = False
+
+    patch_config_class(TransformerConfig)
+    patch_config_class(MLATransformerConfig)
```
[the remark this answers is not in the document]
Comment (tntnnlrw) [reply verl/utils/megatron/router_replay_patch.py]: Fixed in 0a7d3827. `apply_router_replay_patch()` now always patches `TransformerConfig` and only patches `MLATransformerConfig` when it is available. Added a focused regression for the no-MLA Megatron case.

**Comment by tntnnlrw** [comment]
Hi @HollowMan6 @wuxibin89, could either of you take a look when you have bandwidth?

This is a small follow-up around the Megatron router replay config path touched by #6647 and #5884. The fix preserves `enable_routing_replay` through `check_and_construct_configs()` when it filters override kwargs with `hasattr(cls, key)`, and keeps compatibility with older Megatron-LM versions where `MLATransformerConfig` is unavailable.

I addressed the Gemini feedback in `0a7d3827` with conditional MLA import/patching and added a CPU-only focused regression for both config classes and the no-MLA case.

Validation run:
- `pytest tests/utils/megatron/test_router_replay_patch_config.py -q`
- `python -m py_compile verl/utils/megatron/router_replay_patch.py tests/utils/megatron/test_router_replay_patch_config.py`
- `git diff --check HEAD~1..HEAD`

**Review by HollowMan6 — COMMENTED** [review]
I'm a little bit of confused about what this PR is targetting at. Is this fix intended for older mcore versions where `moe_enable_routing_replay` flag hasn't been introduced? https://github.com/NVIDIA/Megatron-LM/commit/db6b895c69b71d2db7f5fd58d3f6de76ed40a619

For the newer mcore version, I think #6647 should have already been addressed.

**Comment by tntnnlrw** [comment]
Thanks @HollowMan6, yes, that is the intended scope.

#6647 already covers the newer mcore path where the official `moe_enable_routing_replay` flag exists. This PR is not trying to replace that. It fixes the existing fallback path in current `main` for the case where the provider does not expose `moe_enable_routing_replay` and verl falls back to `enable_routing_replay`.

The bug is that the fallback key can be silently dropped before config construction. `check_and_construct_configs()` filters override kwargs with `hasattr(cls, key)`, but `enable_routing_replay` was not visible as a class-level attribute on the Megatron config class, so the fallback value could be removed even though `transformer_impl.py` selected it intentionally.

So the goal is:
- newer mcore: keep using `moe_enable_routing_replay` from #6647
- older / fallback path: make the existing `enable_routing_replay` fallback actually survive the converter filter
- MLA config: patch `MLATransformerConfig` as well because its generated dataclass init does not necessarily inherit the patched `TransformerConfig.__init__`

If verl decides to drop this fallback path entirely, then this PR would not be needed. But as long as `main` keeps the fallback branch, this patch makes that branch functional instead of silently no-oping.

**Comment by tntnnlrw** [comment]
Update on CI status:

The core checks relevant to this patch are now passing:
- `cpu_unit_tests`
- `pre-commit (3.12)`
- `sanity (3.10)`
- `type-coverage-check`
- `doc_test (3.10)`
- `check-title`
- `secrets_scan`

Local validation also passed for the focused CPU regression test, ruff, py_compile, and diff whitespace check.

The remaining red/pending jobs appear to be outside this patch path. Several fail during dependency installation because the CI mirror cannot resolve `TransferQueue==0.1.8`, which was introduced on main by #6652, and one fully-async FSDP2 job failed with CUDA OOM. This PR only touches the Megatron router replay fallback config handling plus a CPU-only regression test.

**Comment by tntnnlrw** [comment]
@HollowMan6 Thanks again for checking the fallback scope earlier.

The core checks relevant to this patch are passing now (`cpu_unit_tests`, `pre-commit`, `sanity`, `type-coverage-check`, `doc_test`, title/secrets). `gh pr checks --required` also reports no required checks for this branch.

The remaining red jobs look unrelated to this small Megatron config fallback patch, mainly dependency install issues around `TransferQueue==0.1.8` from current main / #6652 and a few heavy E2E environment failures.

Do you think the current CI signal is sufficient for review, or should I wait for a specific maintainer / workflow rerun before proceeding?

**Comment by tntnnlrw** [comment]
Final CI update:

- 91 checks passed, no pending checks remain.
- `gh pr checks --required` reports no required checks for this branch.
- The remaining 3 failures look unrelated to this patch: one CUDA OOM, one Ascend checkpoint/runner file-write failure, and one NPU/MindSpeed `TransformerConfig` dataclass wrapping failure not involving `router_replay_patch.py`.

The focused CPU regression test and core checks for this Megatron config fallback fix are passing.

**Review by HollowMan6 — APPROVED** [review]

**Review by HollowMan6 — CHANGES_REQUESTED** [review]
Will need rebase, and not sure this is still needed since we have already bumped to new mcore version (0.18.0)

**Comment by HollowMan6** [comment]
Close for now, please feel free to reopen if you still think this is necessary
